src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs

```diff
@@ -516,8 +516,8 @@ private BoundExpression MakeSpanStringTest(BoundExpression input, ConstantValue
                     .Any(static node => node is BoundWhenDecisionDagNode { WhenExpression: { ConstantValue: null } });
 
                 var inputDagTemp = BoundDagTemp.ForOriginalInput(loweredInput);
-                if ((loweredInput.Kind == BoundKind.Local || loweredInput.Kind == BoundKind.Parameter)
-                    && loweredInput.GetRefKind() == RefKind.None &&
+                if (loweredInput.Kind is BoundKind.Local or BoundKind.Parameter &&
+                    loweredInput.GetRefKind() is RefKind.None or RefKind.In &&
```

Comment:
we should test a scenario where a "mutating" member is accessed in the pattern

```cs
void M(in S s)
{
    if (s.Value is 1) { Console.Write(1); }
    if (s.Value is 1) { Console.Write(2); }
}

struct S
{
    int item;
    public int Value => ++item;
}
```

I am pretty sure I would expect implicit copies to be made here and to get output of `12`.

Reply:
See the comment below (line 523). The language and compiler assume that functions called by pattern-matching are not side-effecting.

Reply:
Discussed offline. We agree that when the argument is a readonly reference (i.e. `in`), and the parameter on the method/accessor is a writable reference (i.e. `ref`), then we still need to make sure we copy the value and pass a writable reference to the copy, in order to preserve readonly safety.

I think we should try changing the condition to try and identify/filter out that specific scenario. Unfortunately, since the access could happen anywhere in the DAG, we might have to go ahead and do a DAG traversal (maybe a partial traversal?) to decide if a writable-reference member is ever accessed on the original input.

First step is definitely to just add the test, and particularly the flavor with `s is { Value: 1 }` repeated, to determine whether the current implementation results in a readonly safety violation.

Let me know if this makes sense @ArcadeMode, I'm happy to assist further if I can.

Reply:
I have added the mentioned test cases and  readonly safety appears to be okay.